src/main/java/org/broadinstitute/hellbender/tools/exome/segmentation/ClusteringGenomicHMMSegmenter.java

```diff
@@ -0,0 +1,305 @@
+package org.broadinstitute.hellbender.tools.exome.segmentation;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.commons.math3.analysis.integration.SimpsonIntegrator;
+import org.apache.commons.math3.analysis.integration.UnivariateIntegrator;
+import org.apache.commons.math3.special.Gamma;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.tools.exome.ModeledSegment;
+import org.broadinstitute.hellbender.utils.*;
+import org.broadinstitute.hellbender.utils.hmm.ForwardBackwardAlgorithm;
+import org.broadinstitute.hellbender.utils.hmm.ViterbiAlgorithm;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+/**
+ * Segmentation and parameter learning corresponding to ClusteringGenomicHMM
+ *
+ * @author David Benjamin &lt;davidben@broadinstitute.org&gt;
+ */
+public abstract class ClusteringGenomicHMMSegmenter<T> {
+    protected final Logger logger = LogManager.getLogger(ClusteringGenomicHMM.class);
+
+    private double concentration;
+    protected double[] weights; //one per hidden state
+    protected double[] hiddenStateValues;
+    protected double memoryLength;
+    private boolean parametersHaveBeenLearned = false;
+
+    protected static final int NEUTRAL_VALUE_INDEX = 0;
+
+    private static final int MAX_EM_ITERATIONS = 50;
+    protected final List<T> data;
+    protected final List<SimpleInterval> positions;
+    private final double[] distances;   //distances[n] is the n to n+1 distance
+
+    protected static double NEGLIGIBLE_POSTERIOR_FOR_M_STEP = 0.001;
+
+    //private static final Logger logger = LogManager.getLogger(AlleleFractionSegmenter.class);
+    private static final double MINIMUM_MEMORY_LENGTH = 1;
+    private static final double MAXIMUM_MEMORY_LENGTH = 1e10;
+    private static final double DEFAULT_MEMORY_LENGTH = 5e7;
+
+    //paraeters for pruning unused hidden states
+    final double DISTANCE_TO_NEIGHBOR_TO_BE_CONSIDERED_SPURIOUS = 0.01;  //if a hidden state is this close to another state, it might be false
+    final double MAX_WEIGHT_CONSIDERED_FOR_PRUNING = 0.02;  // never prune a state with greater weight than this
+    final double AUTOMATICALLY_PRUNED_WEIGHT = 5e-4;    //a weight so low it is always pruned
+
+    // (unnormalized) vague gamma prior on concentration
+    private static final Function<Double, Double> PRIOR_ON_CONCENTRATION = alpha -> alpha*Math.exp(-alpha);
+    private static final double MINIMUM_CONCENTRATION = 1e-4;
+    private static final double MAXIMUM_CONCENTRATION = 5;
+    private static final int MAX_INTEGRATION_EVALUATIONS = 1000;
+    private static final UnivariateIntegrator UNIVARIATE_INTEGRATOR = new SimpsonIntegrator(1e-3, 1e-3, 5, 20);
```

Comment:
some magic numbers here.  not necessary to extract, but maybe some comments as to why you chose them?